keps/sig-scheduling/3101-pod-to-node-assignment/README.md

```diff
@@ -0,0 +1,944 @@
+pod-to-node assignment. The most common use case for enforcing 1:1 pod to node
+assignment is HPC workloads (MPI and Spark).
+
+Currently there are two approaches to achieve this:
+
+-   Set resource requests to something high enough to prevent scheduling any
+    other pods but daemonset and static pods. This is tricky to do and error
+    prone, especially for smaller nodes. This will not be possible if node size
+    is not known in advance.
+-   Use pod anti affinity. Users will need to define the same set of labels on
+    all pods that needs to be repelled, whether or not they actually request 1:1
+    pod to node assignment, it is also expensive to calculate especially for
+    cluster autoscaler.
+
+A simple and native way to force 1:1 pod-to-node assignment is therefore needed.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+-   Allow users to force 1:1 pod-to-node assignment for a given pod.
+    -   This is enforced by both scheduler and kubelet.
+    -   This feature is implemented as a filter, and so features associated with
+        filters applies to it like preemption and autoscaling.
+
+### Non-Goals
+
+-   Find the best fit node (a.k.a `Score`) from the nodes that satisfy
+    `pod.spec.CoexistPolicy`
+-   Change node ranking when selecting a candidate for preemption.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Add a new pod-level API that allows users to enforce how a pod can coexist with
+other pods when scheduled on a node. Available policies are:
+
+-   `DaemonsetAndStaticPods`: pod only tolerates daemonset and static pods to be
+    running on the node.
+-   `Any`: pod tolerates any pod type to be running on the node (default)
+
+The API is implemented by the scheduler as a filter plugin and kubelet as an
+admission predicate.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+A user wants a Pod to be scheduled alone on a node to avoid them from contending
+on resources (e.g. claiming all CPU cores available on a node). In other words,
+resource intensive workloads are needed to be scheduled on separate nodes.
+
+Example: As a user, I want to deploy a Spark or MPI job such that only a single
+worker pod is running on each node.
+
+#### Story 2
+
+A DBaaS running natively on Kubernetes which uses 1:1 Pod to Node assignment
```

Comment:
Is the intent here to exclude other pods from that same workload from being colocated?  I.e, the user needs a hard guarantee that pods from workload A are never doubled up, because some resource that kubernetes / linux does not strongly isolate could cause interference?

Same as above, would like a bit more detail.

Reply:
The intent is to have a single workload pod running on the node, in this case it would be a database server pod. But we still want to allow daemonsets to run (for logging, monitoring etc.)